### PR TITLE
[Graphbolt] Support loading heterogeneous attributes in sampling graph.

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by czkkkkkk** [comment]
@dgl-bot

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 1]
```diff
@@ -129,14 +129,41 @@ def preprocess_ondisk_dataset(
                     graph_feature["format"],
                     in_memory=in_memory,
                 )
-                g.ndata[graph_feature["name"]] = node_data
+                if is_homogeneous:
+                    g.ndata[graph_feature["name"]] = node_data
+                else:
+                    g.nodes[graph_feature["name"]] = {
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:135]: is this workable? the canonical way is `g.nodes[NODE_TYPE].data['x'] = xxx`?

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 2]
```diff
@@ -129,14 +129,41 @@ def preprocess_ondisk_dataset(
...
                 edge_data = read_data(
                     os.path.join(dataset_dir, graph_feature["path"]),
                     graph_feature["format"],
                     in_memory=in_memory,
                 )
-                g.edata[graph_feature["name"]] = edge_data
+                if is_homogeneous:
+                    g.edata[graph_feature["name"]] = edge_data
+                else:
+                    g.edata[graph_feature["name"]] = {
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:147]: `g.edges[EDGE_TYPE].data['x'] = xxx`?

**Review by Rhett-Ying — CHANGES_REQUESTED** [review]
please add testcases for this change

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 1 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:135]: I see, I will change to the canonical solution. But my solution is workable as in https://docs.dgl.ai/en/latest/generated/dgl.DGLGraph.ndata.html.

**Comment by czkkkkkk** [comment]
> please add testcases for this change

The changes are workable in TGIF. So do you think it is fine to add unittests in a later PR?

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 1 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:135]: Oh, should be `g.ndata` instead of `g.nodes`. This is not tested in TGIF.

Comment (Rhett-Ying) [reply python/dgl/graphbolt/impl/ondisk_dataset.py]: I'm ok with `g.ndata`.

**Review by Rhett-Ying — APPROVED** [review]

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 3]
```diff
@@ -129,14 +129,41 @@ def preprocess_ondisk_dataset(
...
                 )
-                g.edata[graph_feature["name"]] = edge_data
+                if is_homogeneous:
+                    g.edata[graph_feature["name"]] = edge_data
+                else:
+                    g.edges[etype_str_to_tuple(graph_feature["type"])].data[
+                        graph_feature["name"]
+                    ] = edge_data
+        if not is_homogeneous:
+            # For homogeneous graph, a node/edge feature must cover all
```
Comment (Skeleton003) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:151]: For heterogenous graph?

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 4]
```diff
@@ -129,14 +129,41 @@ def preprocess_ondisk_dataset(
...
+                    + f"Existing types: {existing_types}."
+                    + f"Expected types: {g.ntypes}."
+                )
+            for feat_name, feat_data in g.edata.items():
+                existing_types = set(feat_data.keys())
+                assert existing_types == set(g.canonical_etypes), (
+                    f"Edge feature {feat_name} does not cover all edge types."
+                    + f"Existing types: {existing_types}."
+                    + f"Expected types: {g.etypes}."
+                )
```
Comment (Skeleton003) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:166]: why not put this before line 120?

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 3 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:151]: Yes. Should be `heterogeneous graph`.

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 4 above)
Comment (czkkkkkk) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:166]: This is because we need to collect all features first.

**Pull request merged**